realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -907,66 +959,61 @@ void BaseRealSenseNode::publishFrame(
         height = timage.get_height();
         bpp = timage.get_bytes_per_pixel();
     }
-    auto& image = images[stream];
 
-    if (image.size() != cv::Size(width, height) || image.depth() != _image_formats[bpp])
+    if (image_publishers.find(stream) != image_publishers.end())
     {
-        image.create(height, width, _image_formats[bpp]);
-    }
-    image.data = (uint8_t*)f.get_data();
+        auto &image_publisher = image_publishers.at(stream);
+        cv::Mat image_cv_matrix;
+        if (_rgbd_publisher && 0 != _rgbd_publisher->get_subscription_count())
+        {
+            image_cv_matrix = getCVMatImage(f, images, width, height, bpp, stream);
+        }
+        if (0 != image_publisher->get_subscription_count())
+        {
+            if(image_cv_matrix.empty())
+            {
+                image_cv_matrix = getCVMatImage(f, images, width, height, bpp, stream);
+            }
 
-    if (f.is<rs2::depth_frame>())
-    {
-        image = fix_depth_scale(image, _depth_scaled_image[stream]);
-    }
+            // Prepare image topic to be published
+            // We use UniquePtr for allow intra-process publish when subscribers of that type are available
+            sensor_msgs::msg::Image::UniquePtr img_msg_ptr(new sensor_msgs::msg::Image());
+            fillMessageImage(image_cv_matrix, stream, width, height, bpp, t, img_msg_ptr.get());
+            if (!img_msg_ptr)
+            {
+                ROS_ERROR("sensor image message allocation failed, frame was dropped");
+                return;
+            }
+            
+            // Transfer the unique pointer ownership to the RMW
+            sensor_msgs::msg::Image *msg_address = img_msg_ptr.get();
+            image_publisher->publish(std::move(img_msg_ptr));
 
+            ROS_DEBUG_STREAM(rs2_stream_to_string(f.get_profile().stream_type()) << " stream published, message address: " << std::hex << msg_address);
+        }
+    }
     // if stream is on, and is not a SC stream, publish cameraInfo
     if(shouldPublishCameraInfo(stream) && info_publishers.find(stream) != info_publishers.end())
     {
         auto& info_publisher = info_publishers.at(stream);
-        if(0 != info_publisher->get_subscription_count())
+        if (_rgbd_publisher && 0 != _rgbd_publisher->get_subscription_count())
         {
             auto& cam_info = _camera_info.at(stream);
             if (cam_info.width != width)
             {
                 updateStreamCalibData(f.get_profile().as<rs2::video_stream_profile>());
             }
             cam_info.header.stamp = t;
```

Comment:
We set a field in `cam_info` here, and override it on line 1010.
Why is that?

Reply:
line 999, if there any subscribers to our RGBD topic, we should make sure that cam_info msg is ready, but still not publishing it, since we will do that when the whole message RGBD message is ready (two images, and two cam infos)
*notice, that there might be no subscribers to the depth and color topics, although the user enabled these streams.

in line 1010, we are sure that there are subscribers to info topics of depth/color camera info, so we are publishing the cam info there.
here, we can take the cam_info that was already calculated above (if there any subscriber to RGBD) or calculate it from scratch.

Reply:
I think it's worth to add some comment on lines that there is some assumption for the flow.

Otherwise we can easily change it in the future without knowing that the one who wrote it took in mind some special scenarios.

Currently its hard to understand the flow, so if we postpone the refactor, lets add comments..

Reply:
Sure, will update this.

Reply:
Updated.